Layout fixes for RTL child flows in LTR parents

mbrubeck opened the pull request:
...and vice-versa.  This is not a complete fix for all mixed-direction layout cases, but it fixes enough problems to make some simple test cases pass, like tha attached reftest.

There are FIXME comments for many of the remaining issues.  In particular, this does not yet handle RTL layout of fixed/absolute elements, nor does it completely fix the overconstrained_block reftest.

r? @SimonSapin or @pcwalton

mbrubeck committed:
Layout fixes for RTL child flows in LTR parents

...and vice-versa.  This is not a complete fix for all mixed-direction layout
cases, but it fixes enough problems to make some simple test cases pass, like
tha attached reftest.

There are FIXME comments for many of the remaining issues.  In particular,
this does not yet handle RTL layout of fixed/absolute elements.

pcwalton commented on components/layout/fragment.rs:
```diff
@@ -1974,8 +1975,8 @@ impl Fragment {
-        // FIXME(pcwalton, #2795): Get the real container size.
-        let container_size = Size2D::zero();
+        // FIXME (mbrubeck): Get the real container size, taking vertical writing modes into account.
+        let container_size = Size2D(relative_containing_block_size.inline, relative_containing_block_size.block);
```
Why not use `.to_physical()` here?

mbrubeck replied:
This code doesn't know the relative containing block's writing mode, so it can't call to_physical.  I was going to fix this in a followup, either by adding parameters to thread the writing mode into this function along with the size, or by giving each flow/fragment a direct link to its containing block.

mbrubeck replied:
Actually, we should maybe just store `relative_containing_block_size` in physical coordinates, since I don't think we ever need it in logical coordinates.  I'll try this.

pcwalton commented:
Looks good overall—I have one question.

mbrubeck committed:
Keep track of the containing block writing mode.

This is necessary for correctly converting `relative_containing_block_size`
to physical coordinates.

mbrubeck commented:
> > Why not use .to_physical() here?
> 
> This code doesn't know the relative containing block's writing mode, so it can't call to_physical. I was going to fix this in a followup, either by adding parameters to thread the writing mode into this function along with the size, or by giving each flow/fragment a direct link to its containing block.
>
> Actually, we should maybe just store relative_containing_block_size in physical coordinates, since I don't think we ever need it in logical coordinates. I'll try this.

It turns out we do use the containing block size in both logical and physical coordinates, so we need to pass its writing mode into `stacking_relative_border_box`, which I've done in the new commit here.

mbrubeck commented:
re-r? @pcwalton

pcwalton commented:
Looks good to me, rebase and r=me
